compiler/src/jdk.graal.compiler.libgraal/src/jdk/graal/compiler/libgraal/LibGraalFeature.java

```diff
@@ -0,0 +1,384 @@
+
+        // Mark all the Node classes as allocated so they are available during graph decoding.
+        for (NodeClass<?> nodeClass : encodedSnippets.getSnippetNodeClasses()) {
+            access.registerAsInHeap(nodeClass.getClazz());
+        }
+        HotSpotReplacementsImpl.setEncodedSnippets(encodedSnippets);
+
+        List<ForeignCallSignature> foreignCallSignatures = (List<ForeignCallSignature>) libgraalObjects.get("foreignCallSignatures");
+        HotSpotForeignCallLinkage.Stubs.initStubs(foreignCallSignatures);
+
+        TruffleHostEnvironment.overrideLookup(new LibGraalTruffleHostEnvironmentLookup());
+    }
+
+    /**
+     * Initialization of JVMCI code that needs to be done for JDK versions that do not include
+     * JDK-8346781.
+     */
+    private void doLegacyJVMCIInitialization() {
+        if (!BeforeJDK8346781.VALUE) {
+            return;
+        }
+        try {
+            String rawArch = GraalServices.getSavedProperty("os.arch");
+            String arch = switch (rawArch) {
+                case "x86_64", "amd64" -> "AMD64";
+                case "aarch64" -> "aarch64";
+                case "riscv64" -> "riscv64";
+                default -> throw new GraalError("Unknown or unsupported arch: %s", rawArch);
+            };
+
+            ClassLoader cl = (ClassLoader) libgraalLoader;
+            Field cachedHotSpotJVMCIBackendFactoriesField = ObjectCopier.getField(HotSpotJVMCIRuntime.class, "cachedHotSpotJVMCIBackendFactories");
+            GraalError.guarantee(cachedHotSpotJVMCIBackendFactoriesField.get(null) == null, "Expect cachedHotSpotJVMCIBackendFactories to be null");
+            ServiceLoader<HotSpotJVMCIBackendFactory> load = ServiceLoader.load(HotSpotJVMCIBackendFactory.class, cl);
+            List<HotSpotJVMCIBackendFactory> backendFactories = load.stream()//
+                            .map(ServiceLoader.Provider::get)//
+                            .filter(s -> s.getArchitecture().equals(arch))//
+                            .toList();
+            cachedHotSpotJVMCIBackendFactoriesField.set(null, backendFactories);
+            GraalError.guarantee(backendFactories.size() == 1, "%s", backendFactories);
+
+            var jvmciServiceLocatorCachedLocatorsField = ObjectCopier.getField(JVMCIServiceLocator.class, "cachedLocators");
+            GraalError.guarantee(jvmciServiceLocatorCachedLocatorsField.get(null) == null, "Expect cachedLocators to be null");
+            Iterable<JVMCIServiceLocator> serviceLocators = ServiceLoader.load(JVMCIServiceLocator.class, cl);
+            List<JVMCIServiceLocator> cachedLocators = new ArrayList<>();
+            serviceLocators.forEach(cachedLocators::add);
+            jvmciServiceLocatorCachedLocatorsField.set(null, cachedLocators);
+        } catch (Throwable e) {
+            throw new GraalError(e);
+        }
+    }
+
+    @Override
+    public void duringAnalysis(DuringAnalysisAccess access) {
```

Comment:
Yes, `duringAnalysis` is official API but it's something that we want to get rid of in the long run because `duringAnalysis` implementations run single-threaded, thus having the potential to increase static analysis time significantly.

Using `org.graalvm.nativeimage.hosted.Feature.DuringSetupAccess#registerObjectReplacer`s that just return the original unmodified objects but registers them as UnsafeAccessed as a side-effect was not an option?

Reply:
Done: https://github.com/oracle/graal/pull/10380/commits/9e41cfe0d15cb37f043a4240462f34755a6f7f7e